### PR TITLE
[skip-changelog] Improve the definition of FQBN and explicitly state which characters are allowed

### PR DESCRIPTION
**Pull request by MatteoPologruto**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)
- [ ] `configuration.schema.json` updated if new parameters are added.

## What kind of change does this PR introduce?
Documentation enhancement

## What is the new behavior?
Each field of the `fqbn` is checked to verify if it contains invalid characters.

## Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?
No

### CODE REVIEW

**Review by alessio-perugini — CHANGES_REQUESTED** [review]
I remember that talking with @cmaglie we wanted to enforce this check in the codebase. 
Let's wait for his response if we want to tackle that in the same PR.

File: docs/FAQ.md [hunk 1]
```diff
@@ -9,7 +9,20 @@ When you run [`arduino-cli board list`][arduino cli board list], your board does
...
+key-value pair configuration. Each field accepts letters (`A-Z` or `a-z`), numbers (`0-9`), underscores (`_`),
+dashes(`-`) and dots(`.`). The special character `=` is accepted in the configuration value. For a deeper understanding
+of how FQBN works, you should understand the [Arduino platform specification][0].
+
+## What fields of the FQBN are optional?
+
+The only mandatory field is `BOARD_ID`. When configs are specified, keys cannot be empty while values can. The following
+FQBN examples are valid:
+
+- `arduino::uno`, `:avr:uno`, `::uno`
```
Comment (cmaglie) [inline docs/FAQ.md:23]: I don't understand this piece, AFAIK the mandatory part is the triple `VENDOR:ARCHITECTURE:BOARD_ID` (not just `BOARD_ID`).
The formula above is correct IMHO:
```
VENDOR:ARCHITECTURE:BOARD_ID[:MENU_ID=OPTION_ID[,MENU2_ID=OPTION_ID ...]]
```
from here we can deduce that the optional part is the MENU_ID=OPTION_ID.
Also, the OPTION_ID must match one of the possible value for the corresponding MENU_ID, it cannot be omitted:

```
$ arduino-cli board details -b arduino:avr:mega:cpu=
Error getting board details: Unknown FQBN: getting build properties for board arduino:avr:mega: invalid value '' for option 'cpu'
$ arduino-cli board details -b arduino:avr:mega:cpu
Error getting board details: Invalid FQBN: invalid config option: cpu
$ arduino-cli board details -b arduino:avr:mega:cpu=atmega2560
[...correct output is displayed...]
```

IMHO we should remove the `## What fields of the FQBN are optional?` section entirely.

**Review by cmaglie — CHANGES_REQUESTED** [review]

**Comment by cmaglie** [comment]
> I remember that talking with @cmaglie we wanted to enforce this check in the codebase. Let's wait for his response if we want to tackle that in the same PR.

Yep, it would be nice to start enforcing it.
I don't know how much platforms would break, BTW it seems very unlikely (in case we may relax the specification to include some extra char later).

**Review by cmaglie — APPROVED** [review]

**Review by alessio-perugini — APPROVED** [review]

**Comment by dankeboy36** [comment]
Thanks for the feature. I see `arduin-o:av-r:un-o:a=b=c=d` is valid. What config option and value does it imply? Can you confirm that config option `a` has a value of `"b=c=d"`?

**Comment by MatteoPologruto** [comment]
Hello @dankeboy36! Yes, I am confirming that `arduin-o:av-r:un-o:a=b=c=d` is a valid FQBN and the config key `a` has the value of `b=c=d`.

**Pull request merged**